mozilla_django_oidc/contrib/drf.py

```diff
@@ -0,0 +1,110 @@
+from requests.exceptions import HTTPError
+
+from mozilla_django_oidc.auth import OIDCAuthenticationBackend
+from mozilla_django_oidc.utils import parse_www_authenticate_header
+
+
+def guess_oidc_backend():
+    """
+    Look through AUTH_BACKENDS to find one that is or extends the mozilla
+    OIDC backend class.
+    """
+    backends = [
+        backend for backend in get_backends()
+        if isinstance(backend, OIDCAuthenticationBackend)
+    ]
+    if not backends:
+        raise ImproperlyConfigured('No backends extending OIDCAuthenticationBackend found!')
+    if len(backends) > 1:
+        raise ImproperlyConfigured('More than one OIDCAuthenticationBackend found!')
+    return backends[0]
+
+
+class OIDCAuthentication(authentication.BaseAuthentication):
+    """
+    Provide OpenID authentication for DRF.
+    """
+
+    # used by the authenticate_header method.
+    www_authenticate_realm = 'api'
+
+    def __init__(self, backend=None):
+        self.backend = backend or guess_oidc_backend()
+
+    def authenticate(self, request):
+        """
+        Authenticate the request and return a tuple of (user, token) or None
+        if there was no authentication attempt.
+        """
+        access_token = self.get_access_token(request)
+
+        if not access_token:
+            return None
+
+        try:
+            user = self.backend.get_or_create_user(access_token, None, None)
+        except HTTPError as exc:
+            resp = exc.response
+
+            # if the oidc provider returns 401, it means the token is invalid.
+            # in that case, we want to return the upstream error message (which
+            # we can get from the www-authentication header) in the response.
+            if resp.status_code == 401 and 'www-authenticate' in resp.headers:
+                data = parse_www_authenticate_header(resp.headers['www-authenticate'])
+                raise exceptions.AuthenticationFailed(data['error_description'])
+
+            # for all other http errors, just re-raise the exception.
+            raise
+        except SuspiciousOperation as exc:
+            # trust that the message in the exception is presentable to the user
+            raise exceptions.AuthenticationFailed('Login failed: %s' % exc)
+
+        if not user:
+            msg = 'Login failed: No user found for the given access token.'
+            raise exceptions.AuthenticationFailed(msg)
+
+        return user, access_token
+
+    def get_access_token(self, request):
+        """
+        Get the access token based on a request.
+
+        Returns None if no authentication details were provided. Raises
+        AuthenticationFailed if the token is incorrect.
+        """
+        header = authentication.get_authorization_header(request).decode('ascii')
+        if not header:
+            return None
+
+        auth = header.split()
+
+        if auth[0].lower() != 'bearer':
+            return None
```

Comment:
What is the reasoning behind return ``None`` if not authentication details are provided but raise exception in the rest of the failures?

Reply:
If `None` is returned here, then `authenticate()` also returns `None`, which means DRF won't return a 401 (which would imply that credentials were provided but were incorrect).